src/hotspot/share/opto/mulnode.cpp

```diff
@@ -2051,99 +2053,75 @@ const Type* RotateRightNode::Value(PhaseGVN* phase) const {
-    return false;
-  }
-  shift = shift->uncast();
-  if (shift == nullptr) {
-    return false;
+//------------------------------ Sum & Mask ------------------------------
+
+// Returns a lower bound on the number of trailing zeros in expr, or -1 if the number
+// cannot be determined.
+static jint AndIL_min_trailing_zeros(const PhaseGVN* phase, const Node* expr, BasicType bt) {
+  expr = expr->uncast();
+  const TypeInteger* type = phase->type(expr)->isa_integer(bt);
+  if (type == nullptr) {
+    return -1;
   }
-  if (phase->type(shift)->isa_integer(bt) == nullptr) {
-    return false;
+
+  if (type->is_con()) {
+    jlong con = type->get_con_as_long(bt);
+    return con == 0L ? (type2aelembytes(bt) * BitsPerByte) : count_trailing_zeros(con);
   }
-  BasicType shift_bt = bt;
-  if (bt == T_LONG && shift->Opcode() == Op_ConvI2L) {
+
+  if (expr->Opcode() == Op_ConvI2L) {
+    expr = expr->in(1)->uncast();
     bt = T_INT;
-    Node* val = shift->in(1);
-    if (val == nullptr) {
-      return false;
-    }
-    val = val->uncast();
-    if (val == nullptr) {
-      return false;
-    }
-    if (val->Opcode() == Op_LShiftI) {
-      shift_bt = T_INT;
-      shift = val;
-      if (phase->type(shift)->isa_integer(bt) == nullptr) {
-        return false;
-      }
-    }
+    type = phase->type(expr)->isa_int();
   }
-  if (shift->Opcode() != Op_LShift(shift_bt)) {
-    if (check_reverse &&
-        (mask->Opcode() == Op_LShift(bt) ||
-         (bt == T_LONG && mask->Opcode() == Op_ConvI2L))) {
-      // try it the other way around
-      return AndIL_shift_and_mask_is_always_zero(phase, mask, shift, bt, false);
+
+  if (expr->Opcode() == Op_LShift(bt)) {
+    const TypeInt* rhs_t = phase->type(expr->in(2))->isa_int();
+    if (rhs_t == nullptr || !rhs_t->is_con()) {
+      return -1;
     }
-    return false;
-  }
-  Node* shift2 = shift->in(2);
-  if (shift2 == nullptr) {
-    return false;
+    return rhs_t->get_con() & (type2aelembytes(bt) * BitsPerByte - 1);
   }
-  const Type* shift2_t = phase->type(shift2);
-  if (!shift2_t->isa_int() || !shift2_t->is_int()->is_con()) {
+
+  return -1;
+}
+
+// Checks whether expr is neutral wrt addition under mask, i.e.
+// an expression of the form (AndX (T+expr) mask) can be simplified to (AndX T mask).
+// The X in AndX must be I or L, depending on bt.
+// Specifically, this holds for the following cases,
+// when the shift value N is large enough to zero out
+// all the set positions of the and-mask M:
+//   (AndI (LShiftI _ #N) #M)
+//   (AndL (LShiftL _ #N) #M)
+//   (AndL (ConvI2L (LShiftI _ #N)) #M)
+// including equivalent constant operands:
+//   (AndI (ConI (_ << #N)) #M)
+//   (AndL (ConL (_ << #N)) #M)
+// The M and N values must satisfy ((-1 << N) & M) == 0.
+static bool AndIL_is_zero_element_under_mask(const PhaseGVN* phase, const Node* expr, const Node* mask, BasicType bt) {
+  jint expr_trailing_zeros = AndIL_min_trailing_zeros(phase, expr, bt);
+  if (expr_trailing_zeros < 0) {
```

Comment:
It feels a little strange that the number of trailing zeros could be negative...
That's why I would return 0 if we can prove nothing. It is still clear that we can do nothing here if it is zero, so we can just compare `<= 0`.

Or what was the reason for returning `-1`?